connectors/sources/google_drive.py

```diff
@@ -0,0 +1,638 @@
+                self._google_drive_client.api_call(
+                    resource="about", method="get", fields="kind"
+                )
+            )
+            logger.info("Successfully connected to the Google Drive.")
+        except Exception:
+            logger.exception("Error while connecting to the Google Drive.")
+            raise
+
+    async def validate_config(self):
+        """Validates whether user inputs are valid or not for configuration field.
+
+        Raises:
+            Exception: The format of service account json is invalid.
+        """
+        self.configuration.check_valid()
+
+        try:
+            json.loads(self.configuration["service_account_credentials"])
+        except ValueError as e:
+            raise ConfigurableFieldValueError(
+                "Google Drive service account is not a valid JSON."
+            ) from e
+
+    async def get_drives(self):
+        """Fetch all shared drive (id, name) from Google Drive
+
+        Yields:
+            dict: Shared drive metadata.
+        """
+
+        async for drive in self._google_drive_client.api_call(
+            resource="drives",
+            method="list",
+            fields="nextPageToken,drives(id,name)",
+            full_response=True,
+            pageSize=100,
+        ):
+            yield drive
+
+    async def retrieve_all_drives(self):
+        """Retrieves all shared drives from Google Drive
+
+        Returns:
+            dict: mapping between drive id and its name
+        """
+        drives = {}
+        async for chunk in self.get_drives():
+            drives_chunk = chunk.get("drives", [])
+            for drive in drives_chunk:
+                drives[drive["id"]] = drive["name"]
+
+        return drives
+
+    async def get_folders(self):
+        """Fetch all folders (id, name, parent) from Google Drive
+
+        Yields:
+            dict: Folder metadata.
+        """
+        async for folder in self._google_drive_client.api_call(
+            resource="files",
+            method="list",
+            corpora="allDrives",
+            fields="nextPageToken,files(id,name,parents)",
+            q=f"mimeType='{FOLDER_MIME_TYPE}' and trashed=false",
+            full_response=True,
+            includeItemsFromAllDrives=True,
+            supportsAllDrives=True,
+            pageSize=1000,
+        ):
+            yield folder
+
+    async def retrieve_all_folders(self):
+        """Retrieves all folders from Google Drive
+
+        Returns:
+            dict: mapping between folder id and its (name, parents)
+        """
+        folders = {}
+        async for chunk in self.get_folders():
+            folders_chunk = chunk.get("files", [])
+            for folder in folders_chunk:
+                folders[folder["id"]] = {
+                    "name": folder["name"],
+                    "parents": folder.get("parents", None),
+                }
+
+        return folders
+
+    async def resolve_paths(self):
```

Comment:
Can we live without this cache? How much memory will it occupy for large drives?

Reply:
Takes `0.5MB` of mem for a setup with 15k folders, multiple drives and 85k files in total.

This cache is needed if we want to have the feature parity with Workplace Search. There is no other way to determine the path on drive. See corresponding implementation in: https://github.com/elastic/ent-search/blob/main/connectors/lib/connectors/content_sources/google_drive/extractor.rb#L61 

One reason for keeping this would be the ability to perform search for a given subpath. On the other hand we could just keep track of `drive_name` where a given file exists (that would be much faster as the lookup will just contain drives).

Reply:
The reason Workplace Search built up these paths was to support filtering rules, IIRC. As is, this would be useful with Basic Sync Rules to help accomplish tasks like "filter all documents in the tree under `/unwanted-dir/`". However, Workplace Search didn't have Advanced Sync Rules, and that's really where we'd want to accomplish this kind of task. If we can limit what trees we access at fetch time, we may not need to build up these paths like this. Might be worth checking with Dana and her plans for Advanced Sync Rules with GDrive.

Reply:
From what I understand we can pass a query `q` to as parameter to API call to filter results at fetch time. Here is the documentation for reference:
- https://developers.google.com/drive/api/guides/ref-search-terms 

From what I understand it only supports filtering for shared drive membership, it doesn't support path-based filtering. I will confirm with Dana the right way forward.

Reply:
Customers have often mentioned that for file based systems, it's extremely valuable to restrict the ingested scope by sharing only specific folders out of an entire GDrive tree structure or the opposite. 

Basic rules, especially with the current first match logic, are "born" to be used for this type of filtering so it would be awesome if we could leverage this. If the performance impact is not worth the cost, I'm open to give Advanced Sync Rules this responsibility completely.

Reply:
We decided to keep this logic, the mem footprint is minimal even for large drives (85k files). The blocking part of the function for large drives takes `~0.3s` and this is acceptable.